docs/architecture_overview.md

```diff
@@ -0,0 +1,130 @@
+
+
+
+Each of these classes can be considered themselves as generalizations which can be configured for usage in specific instances. As such, each of these classes provides a list of parameters that are used to configure their behaviors, and subclasses can provide additional parameters for additional specification. No `Blueprint` should only work for one specific tasks, but it should be a blueprint from which many similar tasks can be built.
+
+The classes are all expected to provide an `Args` class, which is a dataclass that we can use as a structured config alongside [Hydra](https://hydra.cc/) to populate configuration. It should also define an `assert_task_args` method, which ensures that given a set of configuration, a run with that particular configuration will not fail to execute for any configuration related reasons.
+
+There's also the abstraction of the `MephistoDB`, which defines the data operations that Mephisto requires to operate properly. If a specific setup requires specialized data handling, any class implementing this interface could stand in for it. 
+
+## Blueprints
+The [blueprints](https://github.com/facebookresearch/Mephisto/tree/master/mephisto/server/blueprints) contain all of the related code required to set up a task run. Blueprints may follow a hierarchical structure, in that general functionality can be written into abstract blueprints (which are powerful but perhaps hard to configure) and then downstream blueprints may have more configuration control but less breadth. Much of this can actually require significant overhead, so we've created additional abstract classes that a blueprint must link an implementation for. These are listed below:
+### BlueprintArgs
+These define the specific arguments for configuring a blueprint. Some base arguments relevant to all blueprints are provided already, and classes that override `BlueprintArgs` can add anything else they want.
+### SharedTaskState
+Some tasks require additional state that cannot be considered static configuration, such as validation functions for accurate completion of a task, or a counter for the number of tasks completed in a run so far, or any other python state that multiple different assignments may be interested in accessing or altering. This class is for the edge cases that need something more than static configuration.
```

Comment:
Perhaps it's worth mentioning here if this is sort of a singleton state that all tasks have access to, or does each assignment/unit get it's own instance?